satellite/satellitedb/usercredits.go

```diff
@@ -55,21 +55,41 @@ func (c *usercredits) GetCreditUsage(ctx context.Context, userID uuid.UUID, expi
 }
 
 // Create insert a new record of user credit
-func (c *usercredits) Create(ctx context.Context, userCredit console.UserCredit) (*console.UserCredit, error) {
-	credit, err := c.db.Create_UserCredit(ctx,
-		dbx.UserCredit_UserId(userCredit.UserID[:]),
-		dbx.UserCredit_OfferId(userCredit.OfferID),
-		dbx.UserCredit_CreditsEarnedInCents(userCredit.CreditsEarned.Cents()),
-		dbx.UserCredit_ExpiresAt(userCredit.ExpiresAt),
-		dbx.UserCredit_Create_Fields{
-			ReferredBy: dbx.UserCredit_ReferredBy(userCredit.ReferredBy[:]),
-		},
-	)
+func (c *usercredits) Create(ctx context.Context, userCredit console.UserCredit, redeemableCap int) error {
+	var statement string
+
+	switch t := c.db.Driver().(type) {
+	case *sqlite3.SQLiteDriver:
+		statement = `
+			INSERT INTO user_credits (user_id, offer_id, credits_earned_in_cents, credits_used_in_cents, expires_at, referred_by, created_at)
+				SELECT * FROM (VALUES (?, ?, ?, 0, ?, ?, time('now'))) AS v
+					WHERE (SELECT COUNT(offer_id) FROM user_credits WHERE offer_id = ? ) < ?;
+		`
+	case *pq.Driver:
+		statement = `
+			INSERT INTO user_credits (user_id, offer_id, credits_earned_in_cents, credits_used_in_cents, expires_at, referred_by, created_at)
+				SELECT * FROM (VALUES (?::bytea, ?::int, ?::int, 0, ?::timestamp, ?::bytea, now())) AS v
+					WHERE (SELECT COUNT(offer_id) FROM user_credits WHERE offer_id = ? ) < ?;
+		`
+	default:
+		return errs.New("Unsupported database %t", t)
+	}
+
+	result, err := c.db.DB.ExecContext(ctx, c.db.Rebind(statement), userCredit.UserID[:], userCredit.OfferID, userCredit.CreditsEarned.Cents(), userCredit.ExpiresAt, userCredit.ReferredBy[:], userCredit.OfferID, redeemableCap)
 	if err != nil {
-		return nil, errs.Wrap(err)
+		return errs.Wrap(err)
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		return errs.Wrap(err)
 	}
 
-	return convertDBCredit(credit)
+	if rows != 1 {
+		return errs.New("Create user credit failed")
```

Comment:
Should this error message say anything about why, like "this offer was already redeemed the maximum number of times"? Or is that not something we want to expose